src/basic/chattr-util.c

```diff
@@ -172,3 +173,61 @@ int read_attr_at(int dir_fd, const char *path, unsigned *ret) {
 
         return read_attr_fd(fd, ret);
 }
+
+int get_proj_id(int fd, uint32_t *ret) {
+        struct fsxattr attrs;
+
+        assert(ret);
+
+        if (ioctl(fd, FS_IOC_FSGETXATTR, &attrs) < 0)
+                return -errno;
+
+        *ret = attrs.fsx_projid;
+
+        return 0;
+}
+
+int set_proj_id(int inode_fd, uint32_t proj_id) {
+        struct fsxattr attrs;
+        _cleanup_close_ int fd = -EBADF;
+
+        fd = fd_reopen(inode_fd, O_RDONLY|O_CLOEXEC);
```

Comment:
might be a good case to optimize things as:

```c
int set_proj_id(int fd, uint32_t proj_id) {
        _cleanup_close_ int fd_reopened = -EBADF;
        fd = fd_reopen_condition(fd, 0, O_PATH, &fd_reopened);
        if (fd < 0)
                return fd;
        …
```

Reply:
and please make set_proj_id() and get_proj_id() symmetric: i.e. do this reopening business in both